src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadBuildManager.cs

```diff
@@ -13,44 +14,54 @@ internal sealed class ProjectHotReloadBuildManager(
     IVsService<SVsSolutionBuildManager, IVsSolutionBuildManager2> solutionBuildManagerService) : IProjectHotReloadBuildManager
 {
     private IVsSolutionBuildManager2? _vsSolutionBuildManager2;
+    private readonly ReentrantSemaphore _semaphore = ReentrantSemaphore.Create(
+        initialCount: 1,
+        joinableTaskContext: threadingService.JoinableTaskContext.Context,
+        mode: ReentrantSemaphore.ReentrancyMode.Freeform);
```

Comment:
I think we want this to be: NotRecognized

Reply:
> `NotRecognized` means you'll deadlock if a semaphore-holder tries to enter it again.

There's no nested call for `BuildProjectAsync` on the same thread, would that still be a deadlock if using `NotRecognized`?

Reply:
The nested calls don't have to be on the same thread. They just have to be made in the same or a derived execution context. 
`NotRecognized` will deadlock someone who holds a semaphore and is asking again to enter one (but won't release the original one while waiting).